pkg/rtc/transport.go

```diff
@@ -337,12 +352,37 @@ func (t *PCTransport) Close() {
 
 func (t *PCTransport) SetRemoteDescription(sd webrtc.SessionDescription) error {
 	t.lock.Lock()
-	defer t.lock.Unlock()
+
+	var (
+		iceCredential   string
+		offerRestartICE bool
+	)
+	if sd.Type == webrtc.SDPTypeOffer {
+		var err error
+		iceCredential, offerRestartICE, err = t.isRemoteOfferRestartICE(sd)
+		if err != nil {
+			t.Logger().Errorw("check remote offer restart ice failed", err)
+			t.lock.Unlock()
+			return err
+		}
+	}
+
+	if offerRestartICE && t.pc.ICEGatheringState() == webrtc.ICEGatheringStateGathering {
+		t.Logger().Debugw("remote offer restart ice while ice gathering")
+		t.pendingRestartIceOffer = &sd
+		t.lock.Unlock()
+		return nil
+	}
 
 	if err := t.pc.SetRemoteDescription(sd); err != nil {
+		t.lock.Unlock()
 		return err
 	}
 
+	if t.currentOfferIceCredential == "" || offerRestartICE {
```

Comment:
should these checks all move into the `sd.Type == webrtc.SDPTypeOffer` block?

Reply:
we only save `iceCredential` if SetRemoteDescription success.